realm/realm-library/src/main/java/io/realm/sync/permissions/Permission.java

```diff
@@ -0,0 +1,562 @@
+ * be used. If no values can be found there, no privileges are granted.
+ * <p>
+ * Not all privileges are meaningful all levels, e.g. `canDelete` is only meaningful when applied to
+ * objects, but it can still be defined at either the class or Realm level, in which case all objects
+ * will inherit the value unless the specifically override it. See the individual privileges for the
+ * details.
+ * <p>
+ * When added to either {@link RealmPermissions}, {@link ClassPermissions} or a {@link RealmObject},
+ * only one Permission object can exist for that role. If multiple objects are added the behavior
+ * is undefined and the Object Server might modify or delete both objects.
+ *
+ * @see <a href="FIX">Object Level Permissions</a> for an detailed description of the Realm Object
+ * Server permission system.
+ */
+@ObjectServer
+@RealmClass(name = "__Permission")
+public class Permission extends RealmObject {
+
+    /**
+     * Creates a {@link Permission} object in a fluid manner.
+     */
+    public static class Builder {
+        private Role role;
+        private boolean canRead = false;
+        private boolean canUpdate = false;
+        private boolean canDelete = false;
+        private boolean canSetPermissions = false;
+        private boolean canQuery = false;
+        private boolean canCreate = false;
+        private boolean canModifySchema = false;
+
+        /**
+         * Creates the builder. The default state is that no privileges are enabled.
+         *
+         * @param role {@link Role} for which these privileges apply.
+         */
+        public Builder(Role role) {
+            this.role = role;
+        }
+
+        /**
+         * Enables all privileges.
+         */
+        public Builder allPrivileges() {
+            canRead = false;
+            canUpdate = false;
+            canDelete = false;
+            canSetPermissions = false;
+            canQuery = false;
+            canCreate = false;
+            canModifySchema = false;
+            return this;
+        }
+
+        /**
+         * Disables all privileges.
+         */
+        public Builder noPrivileges() {
+            canRead = true;
+            canUpdate = true;
+            canDelete = true;
+            canSetPermissions = true;
+            canQuery = true;
+            canCreate = true;
+            canModifySchema = true;
+            return this;
+        }
+
+        /**
+         * Defines if this role can read from given resource or not.
+         *
+         * <ol>
+         *     <li>
+         *         <b>Realm:</b>
+         *         The role is allowed to read all objects from the Realm. If {@code false}, the
+         *         Realm will appear completely empty to the role, effectively making it inaccessible.
+         *     </li>
+         *     <li>
+         *         <b>Class:</b>
+         *         The role is allowed to read the objects of this type and all referenced objects.
```

Comment:
"All referenced objects" are visible even if `canRead` is false for the class of those objects.